docs(aio): Clarify use of negated class binding

jfahrenkrug opened the pull request:
When the `isSpecial` value is falsy, the `special` CSS class class was being set. This is confusing and the opposite of what a new user might expect. This commit changes the example to set a `not-so-special` CSS class when `isSpecial` is falsy.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[X] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?
In the current documentation, the `special` CSS class is being set when `isSpecial` is false.

## What is the new behavior?
A `not-so-special` CSS class is being set when `isSpecial` is false.

## Does this PR introduce a breaking change?
```
[ ] Yes
[X] No
```

jfahrenkrug committed:
docs(aio): Clarify use of negated class binding

When the `isSpecial` value is falsy, the `special` CSS class was being set. 
This is confusing and the opposite of what  might be expected. 
This commit changes the example to set a `not-so-special` CSS class when `isSpecial` is falsy.

gkalpak commented on aio/content/examples/template-syntax/src/app/app.component.html:
```diff
@@ -337,9 +337,9 @@
-<!-- binding to `class.special` trumps the class attribute -->
-<div class="special"
-     [class.special]="!isSpecial">This one is not so special</div>
+<!-- binding to `class.not-so-special` trumps the class attribute -->
```
This is a little confusing too, because [the guide][1] only refers to the `special` class:

> The following examples show how to add and remove the application's "special" class with class bindings.

Consider updating the wording there too.

##
Not directly related to this PR, but the [e2e tests][2] could use some love. 

[1]: https://github.com/jfahrenkrug/angular/blob/7b29c5586671b24339099cef78e534c95294cbc2/aio/content/guide/template-syntax.md#class-binding
[2]: https://github.com/jfahrenkrug/angular/blob/7b29c5586671b24339099cef78e534c95294cbc2/aio/content/examples/template-syntax/e2e-spec.ts

jfahrenkrug replied:
Great point! You know what, I think it would be even clearer if we did this:

    <!-- toggle the "special" class on/off with a property -->
    <div [class.special]="isSpecial">The class binding is special</div>

    <!-- binding to `class.special` trumps the class attribute -->
    <div class="special"
         [class.special]="isSpecial">The class binding is special</div>

What do you think about that? We leave the negation completely out of this example because it only complicates things. In this section we are talking about class binding and the second code example should show how the class.special binding trumps the class attribute. 
The fact that you *can* negate values is mentioned multiple times throughout the document, so no need to use it here.

gkalpak replied:
TBH, I don't think `<div class="special" [class.special]="isSpecial">` clearly conveys that the `special` class will be removed if `isSpecial` is false, because in all previous examples it is implied that `isSpecial` is true.

I liked your version (`<div class="not-special" [class.not-special]="!isSpecial">`) better. You could just change the guide wording to:

```diff
 The following examples show how to add and remove
-the application's "special" class
+classes
 with class bindings.

(cc @wardbell)

jfahrenkrug replied:
OK good idea! I've made the change.

jfahrenkrug committed:
docs(aio): Clarify use of negated class binding

In the template syntax guide, only generically refer to adding and
removing classes instead of mentioning the specific "special" class.

gkalpak commented on aio/content/examples/template-syntax/src/app/app.component.html:
```diff
@@ -337,9 +337,9 @@
-     [class.special]="!isSpecial">This one is not so special</div>
+<!-- binding to `class.not-so-special` trumps the class attribute -->
+<div class="not-so-special"
+     [class.not-so-special]="!isSpecial">This one is not so special</div>
```
The text content is now confusing. (It is still special, that's why the `not-so-special` class is removed.)
How about changing it to: _This one is special too_

jfahrenkrug replied:
Are you sure? We are binding to the *negated* value of `isSpecial`. So when `isSpecial` is *false*, the `not-so-special` CSS class will be added. So that means that when the `not-so-special` CSS class is present on that div, the content of "This one is not so special" is correct. Or am I way off?

gkalpak replied:
It is implied that `isSpecial` is true (thus `!isSpecial` is false). What this examples is trying to demonstrate is that even if the class attribute already contains a class (i.e. `class="not-so-special"`), a falsy class binding expression (i.e. `[class.not-so-special]="!isSpecial"`) will remove the class anyway. So, the `not-so-special` class will be removed, because _this one is special too_ (a.ka.a not not-so-special).

Does this make any sense? 😇

jfahrenkrug replied:
@gkalpak a-HA! I wasn't aware that `isSpecial` was implied to be true. My thinking was along the lines: "The content of the div will only show when the class is applied." Meaning, "This one is not so special" will only show when the `not-so-special` class is applied.

You know what? I think the fact that this whole thing is so confusing even to the both of us might indicate that it should be taken out of the guide at that point. I think we would not lose anything by taking a negated class binding example out at that point in the guide. It does more harm than good. What do you think?

gkalpak replied:
This example is trying to show something that is neither obvious nor shown in another example. I think there is merit in leaving it in the guide. I am sure we can find a way to make it clear :grin:

petebacondarwin commented:
Are we running round in circles a bit here?

jfahrenkrug commented:
@petebacondarwin @gkalpak Yes, it seems so :) I'm pretty new to Angular and went through the complete guide on angular.io. Now that I'm actually using Angular this little fix seems insignificant tbh. There seem to be much more pressing areas of the guide that seem to have fallen a little behind the current state of Angular, like deployment/build options. Webpack is the way to go now, but the guide still talks about rollup and system.js. 

But that's off-topic :)

I'm fine with whatever you suggest @gkalpak, as long as it ends up being clearer than it currently is ;-)

petebacondarwin commented:
@jfahrenkrug you'll be interested to see we are in the process of migrating all the examples to CLI

gkalpak commented:
@jfahrenkrug, making [this change][1] (along with the rest of the changes you've made) looks cleaner to me (but I am not a technical writer :grin:).
<sub>You are right that are other parts of the docs that need more attention and that's where we are focusing our efforts on. But every bit of improvement in any area <is more than welcome. Thx for stepping up 👍</sub>

[1]: https://github.com/angular/angular/pull/19036/files#r137373980

jfahrenkrug commented:
@petebacondarwin @gkalpak Sorry for the late reply! Thanks both of you for your hard work on this and for your kind words! @gkalpak the change you've pointed out is fine with me :) What would be the next steps for me?

gkalpak commented:
> What would be the next steps for me?

Making [the change][1] :smiley:
Then we can get this merged ✨  

[1]: https://github.com/angular/angular/pull/19036/files#r137373980

kapunahelewong commented:
@gkalpak From the writing end, LGTM. It's clearer IMO. Thank you for this @jfahrenkrug :)

jfahrenkrug commented:
@kapunahelewong Thank you :) I'm sorry I haven't replied in a while :-/

kapunahelewong commented:
@jfahrenkrug Ah, me too. Thank you for your kind patience.

gkalpak commented:
https://github.com/angular/angular/pull/19036/files#r137373980 is still not addressed 😕

jfahrenkrug commented:
@brandonroberts Hi Brandon, thanks for checking. I think we can abandon it. It was a minor thing and it's been a while since I've looked at this. Thanks!

brandonroberts commented:
If you see something else, feel free to open another PR.
